File: ts/install/file.ts
import { join } from 'path'
import { downloadBinary } from '../download'
import { getInstallDir } from '../env'
import {
  displayOutput,
  endsWithExe,
  getBinName,
  getFilename,
  guessName,
  installOutputFiles,
  nameNoExt,
  showSuccess,
} from '../tool'
import { Output } from '../type'
import { getLocalTarget, Os, targetGetOs } from 'guess-target'
import { DefaultMode } from '../const'

export type FileInstall = {
  url: string
  name?: string
}

export async function fileInstall(
  downloadUrl: string,
  name?: string,
  dir?: string,
): Promise<Output> {
  let installDir: string = getInstallDir()
  if (dir) {
    if (dir.includes('/') || dir.includes('\\')) {
      installDir = dir
    } else {
      installDir = join(installDir, dir).replaceAll('\\', '/')
    }
  }
  const filename = name ?? getFilename(downloadUrl)
  const localTarget = getLocalTarget()
  if (
    endsWithExe(downloadUrl) &&
    localTarget.some((i) => targetGetOs(i) !== Os.Windows)
  ) return {}
  const binName = name ?? guessName(nameNoExt(filename))?.name ??
    nameNoExt(filename)
  const originPath = downloadUrl.split('/').at(-1)!
  const isDir = false
  const buffer = new Uint8Array(await downloadBinary(downloadUrl))
  const installPath = join(installDir, getBinName(binName)).replaceAll(
    '\\',
    '/',
  )
  const size = buffer.byteLength
  const files = [{
    size,
    mode: DefaultMode,
    downloadUrl,
    installPath,
    installDir,
    originPath,
    isDir,
    buffer,
  }]
  const output: Output = {
    [downloadUrl]: {
      installDir,
      files,
    },
  }
  installOutputFiles(files)
  showSuccess()
  console.log(displayOutput(output))
  return output
}

File: src/install/file.rs
use crate::InstallConfig;
use crate::download::download_binary;
use crate::env::get_install_dir;
use crate::tool::{
    check_disk_space, display_output, ends_with_exe, expand_path, get_bin_name, get_filename,
    install_output_files, path_to_str,
};
use crate::types::{Output, OutputFile, OutputItem};
use crate::error::Result;
use guess_target::Os;

pub(crate) async fn install_from_single_file(
    url: &str,
    name: &str,
    config: &InstallConfig,
) -> Result<Output> {
    let local_target = config.get_local_target();
    if ends_with_exe(url) && local_target.iter().any(|t| t.os() != Os::Windows) {
        return Ok(Output::new());
    }
    let filename = get_filename(url);
    // Callers pass the full source filename (e.g. `cli.ts`, `run.sh`, `bun`),
    // so known extensions are preserved. get_bin_name only appends a platform
    // extension (.exe) to bare names.
    let bin = if std::fs::exists(url).unwrap_or(false) {
        std::fs::read(url)?
    } else {
        download_binary(url, config.retry, config.timeout).await?
    };
    install_from_buffer(bin, &filename, name, url, config)
}

/// Install an already-downloaded (or decompressed) in-memory file.
///
/// * `origin_path` - the source file name (used for executable detection), e.g. `tool.exe`.
/// * `name` - the base name used for the installed file (platform extension applied).
/// * `key` - the key stored in the returned `Output` map (typically the source URL).
pub(crate) fn install_from_buffer(
    buffer: Vec<u8>,
    origin_path: &str,
    name: &str,
    key: &str,
    config: &InstallConfig,
) -> Result<Output> {
    let mut install_dir = get_install_dir()?;
    let mut output = Output::new();

    if let Some(target_dir) = &config.dir {
        if target_dir.contains("/") || target_dir.contains("\\") {
            install_dir = expand_path(target_dir).into();
        } else {
            install_dir.push(target_dir);
        }
    }

    let mut install_path = install_dir.clone();
    install_path.push(get_bin_name(name));
    let install_path = path_to_str(&install_path);
    let size = buffer.len() as u64;
    let mut files = vec![OutputFile {
        mode: None,
        size,
        origin_path: origin_path.to_string(),
        is_dir: false,
        install_path,
        buffer,
    }];
    check_disk_space(&files, &install_dir)?;
    install_output_files(&mut files, config)?;
    let item = OutputItem {
        install_dir: path_to_str(&install_dir),
        files,
    };
    output.insert(key.to_string(), item);

    if !config.quiet {
        println!("{}", display_output(&output, config));
    }
    Ok(output)
}
